private/buf/buflsp/completion.go

```diff
@@ -0,0 +1,533 @@
+// This includes qualified names with dots (e.g., "buf.validate.").
+// Returns empty string if no valid prefix is found.
+func extractIdentifierPrefix(text string, offset int) string {
+	if offset <= 0 || offset > len(text) {
+		return ""
+	}
+
+	// Walk backwards from the cursor position.
+	start := offset
+	for start > 0 {
+		// Get the byte before current position.
+		b := text[start-1]
+
+		// Check if it's a valid identifier character.
+		// Valid: letters, digits, underscore, dot
+		if isIdentifierChar(b) {
+			start--
+			continue
+		}
+
+		// Hit a non-identifier character, stop.
+		break
+	}
+
+	return text[start:offset]
+}
+
+// isIdentifierChar checks if a byte is a valid identifier character.
+// Valid characters: a-z, A-Z, 0-9, _, .
+func isIdentifierChar(b byte) bool {
+	return (b >= 'a' && b <= 'z') ||
+		(b >= 'A' && b <= 'Z') ||
+		(b >= '0' && b <= '9') ||
+		b == '_' ||
+		b == '.'
+}
+
+// isInNamePosition checks if the cursor offset is in a position where a unique identifier
+// is being defined (like a message name, field name, etc.). We should not provide completions
+// for these positions.
+func isInNamePosition(decl ast.DeclAny, offset int) bool {
+	if decl.IsZero() {
+		return false
+	}
+
+	switch decl.Kind() {
+	case ast.DeclKindDef:
+		def := decl.AsDef()
+		// Check if cursor is in the name of the definition.
+		if !def.Name().IsZero() {
+			nameSpan := def.Name().Span()
+			if offset >= nameSpan.Start && offset <= nameSpan.End {
+				return true
+			}
+		}
+
+		// For fields, also check the field name (not the type).
+		switch def.Classify() {
+		case ast.DefKindField:
+			field := def.AsField()
+			// Check if cursor is within the field name span.
+			if !field.Name.IsZero() {
+				nameSpan := field.Name.Span()
+				if offset >= nameSpan.Start && offset <= nameSpan.End {
+					return true
+				}
+			}
+			// Check if cursor is after the type (where the field name should be).
+			// This handles the case where the user has typed the type but not yet
+			// the field name (e.g., "buf.validate.Rule _").
+			if !field.Type.IsZero() {
+				typeSpan := field.Type.Span()
+				// If cursor is after the type ends, we're in the field name position.
+				if offset > typeSpan.End {
+					return true
+				}
+			}
+		case ast.DefKindEnumValue:
+			enumValue := def.AsEnumValue()
+			if !enumValue.Name.IsZero() {
+				nameSpan := enumValue.Name.Span()
+				if offset >= nameSpan.Start && offset <= nameSpan.End {
+					return true
+				}
+			}
+		case ast.DefKindMethod:
+			method := def.AsMethod()
+			if !method.Name.IsZero() {
+				nameSpan := method.Name.Span()
+				if offset >= nameSpan.Start && offset <= nameSpan.End {
+					return true
+				}
+			}
+		}
+
+	case ast.DeclKindSyntax:
+		// Cursor in syntax declaration - likely in the syntax value, not a good place for completions.
+		return true
+
+	case ast.DeclKindPackage:
+		// Cursor in package declaration - package name is unique.
```

Comment:
Not true: multiple files can have the same package. We should suggest a name based on the file's path or something like that. You can have a TODO.

Reply:
Fixed.